### PR TITLE
Add flags to enable shared workspace for compilation

### PR DESCRIPTION
**Pull request by kel-certora**
## 🚀 Pull Request Overview

- **Related Issues/Tickets**: https://certora.atlassian.net/browse/CERT-9988
- **Type of Change** (bugfix, feature, refactor, etc.): Feature, performance improvement.
- **Breaking Changes?**: No. All changes are guarded behind flags that are neither required nor active by default.

## Summary

Adds three optional inputs that control how configurations are isolated, scheduled, and cleaned up during a run. All default to false, so behavior is unchanged unless a workflow opts in:

Input | Env var | Default | Effect when true
-- | -- | -- | --
use-workspace-dir | CERTORA_USE_WORKSPACE_DIR | false | Run every configuration directly in the working directory instead of copying it to a /tmp/<hash> directory per configuration.
run-serially | CERTORA_RUN_SERIALLY | false | Submit configurations one at a time instead of concurrently.

This helps bringing memory usage, runtime, and disk usage for Solana project down.

## The Problem this Fixes

The Solana prover invokes cargo to build the project, and cargo maintains build cache local to the project directory. When the project is copied to a separate temporary folder per config file, that cache cannot be reused: each copy builds from scratch. On jobs with many confs and large builds, these concurrent isolated builds can lead to long runtimes or OOM errors on the runner.

Enabling the new `use-workspace-dir` runs all confs in the same workspace instead, so they share one build cache, reducing total runtime and memory usage. 

Note that Cargo locks the build directory, so only one compilation runs at a time per workspace; compilation is never concurrent on the same workspace.
However, all confs still write to the same output path. When different conf files enable different cargo features, a later build can overwrite an earlier conf's artifacts before they are submitted, so the wrong binary gets verified. The `run-serially` flag prevents this by letting each conf finish building and submitting before the next starts. A warning is issued in the logs when `use-workspace-dir` is enabled but `run-serially` is not.

~~Finally, the `cleanup-internal-dirs` is needed for larger codebases whose build artifacts in the `.certora_internal` directory fill up the storage space of the runner, leading to `out of device memory` errors.~~

### Comparison to existing v2 runner on Klend

The original problem explained above surfaced on KLend, as explained in this [ticket](https://certora.atlassian.net/browse/CERT-9988). The existing v2 runner could handle at most 7 confs per job without OOM errors; and even with just 7 confs, the job took roughly one hour to complete (unfortunately, the original runs linked in the ticket have been pruned).  When enabling the above flags, the same 7 conf job finishes in 18 minutes ([check run here](https://github.com/Certora/klend-audit/actions/runs/28022435897/job/82941824680?pr=52)).

## Changes Summary

- `action.yml`: 3 new inputs, wired into the Certora Run step's env (mirrors inputs).
- `scripts/run-certora.sh`: gate the copy loop behind use-workspace-dir; run jobs serially (capturing exit codes inline) vs concurrently behind run-serially; ~~per-job cleanup in serial mode plus an end-of-run sweep in concurrent mode;~~ emit a race-condition warning.
- `README.md`: document the three inputs.

---

### 📜 Tests Checklist

Before submitting this PR, ensure that all tests pass **and** meet the following conditions:

| Category                  | Status                          | Compilation Comment | Results Review |
|---------------------------|----------------------------------|----------------------|----------------|
| **fail_to_start**         | ❌ Failed with compilation error | 🟢 Yes               | 🔴 No          |
| **violated_rules**        | ✅ Passed, with violations found | 🔴 No                | 🟢 Yes         |
| **verified_rules**        | ✅ Passed without violations     | 🔴 No                | 🟢 Yes         |
| **solana_violated_rules** | ✅ Passed, with violations found | 🔴 No                | 🟢 Yes         |
| **sui_verified_rules**    | ✅ Passed without violations     | 🔴 No                | 🟢 Yes         |

> Please verify that your changes meet the above conditions for each category of tests.  
> If something doesn't match, investigate before submitting the PR.

---

### 📋 Additional Notes (Optional)

- Any risks, edge cases, or implementation notes for reviewers:

### CODE REVIEW

File: scripts/run-certora.sh [hunk 1]
```diff
@@ -172,6 +202,13 @@ for i in "${!pids[@]}"; do
   fi
 done
 
+# In concurrent mode we cannot delete mid-run, so sweep the workspace build
+# directory now that every job has been submitted.
+# Frees disk for later certora-run steps that share the same disk.
+if [[ "$CERTORA_CLEANUP_INTERNAL_DIRS" == "true" && "$CERTORA_RUN_SERIALLY" != "true" ]]; then
+  rm -rf "$current_dir/.certora_internal"
+fi
+
```
Comment (H00N24) [inline scripts/run-certora.sh]: This will only work for `CERTORA_USE_WORKSPACE_DIR` as it's cleaning only a single directory (not all hashed configurations). Do we want to limit it to this scope only?

Also, if space is the problem we can use something like:
```
      - name: Free disk space
        uses: jlumbroso/free-disk-space@f68fdb76e2ea636224182cfb7377ff9a1708f9b8 # v1.3.0
        with:
          tool-cache: true
          android: true
          dotnet: true
          haskell: true
          large-packages: true
          swap-storage: true
```
This will clear quite a bit of space on GH workers (>>5GB).

**Review by H00N24 — CHANGES_REQUESTED** [review]

File: scripts/run-certora.sh (hunk 1 above)
Comment (kel-certora) [reply scripts/run-certora.sh]: You are right, that was an oversight. Let me check if the Free disk space thing works and if so, I think this flag can be removed completely.

Comment (kel-certora) [reply scripts/run-certora.sh]: Wasn't needed, deleted.

**Review by H00N24 — APPROVED** [review]
